### ISSUE TITLE
Running out of keys

### ISSUE THREAD

**Issue by schrum2** [schrum2/MM-NEAT#389]
This seems to be a real possibility in levels from the original game if you go through doors in the wrong order. I don't know if it is a problem in the generated dungeons yet, but hopefully it is not. Still, we need some fix before the study.

Maybe we could make keys randomly drop in rooms when enemies respawn. However, I want to restrict this. For starters, we should only do this if the player has 0 keys. Even better would be to only do this if we know the player is stuck, meaning the player has 0 keys, AND we know there is a locked door between the player and the goal, but this would require running A* during play, which I would prefer not to do.

So, for starters, let's just drop keys when all enemies are killed, but ONLY if the player has 0 keys, AND the enemies have respawned in the room. The reason for the respawn requirement is that it must mean the player is backtracking, which implies the player might be stuck, which could happen if the player runs out of keys. It's not perfect, but could work.

**Comment by jakeryang** [comment]
>  I don't know if it is a problem in the generated dungeons yet, but hopefully it is not

This should not be a problem, there will always be a 1-to-1 ratio of keys to doors in the generated dungeon

**Comment by jakeryang** [comment]
I think we can assume the player is stuck if they have 0 keys, the dungeon has 0 keys left, and the dungeon still has locked doors. We can keep track of how many keys and locked doors are left in the dungeon instead of having to search the entirety of the dungeon. 

I had a way of handling this in the past, but I removed it since you stated that you saw keys in rooms where they should not have been. I randomly placed keys in the dungeon until they matched with how many locked doors were placed, while not placing a key twice in the same room.

These ways are both easy to implement. But I think the first one would be the best so we can know for sure that a player will get a necessary key. With the second one, the V-UV algorithm may skip a couple of keys leaving them blocked off.

**Comment by schrum2** [comment]
The problem is that the user could be stuck even if there are keys left in the dungeon, because those keys are behind locked doors. So, simply checking the count will be imperfect.

Whatever we do, there needs to be a failsafe so that users will always eventually have a chance at a key ... even if we vet all levels from the study in advance, we can't be certain they will play the way we choose to, so they could get stuck. 

How about this: we track the number of unlocked passages (which includes bomb-able walls and puzzle doors) that the player has seen but not used. We can also track the number of rooms in which a key would have dropped if the player had killed the enemies (can't assume they will). We only allow keys to randomly drop from enemies if the following are true: player has 0 keys, player has traversed all unlocked passages he/she has seen (been in the same room as), and there are no rooms where enemies would have dropped keys if only the player had killed them.

Basically, force the player to explore before attempting to fix the issue.

**Comment by schrum2** [comment]
This problem was only ever witnessed in a dungeon from the original game. Properly designed graph grammars should prevent a lack of keys in those dungeons. There are other issues with keys, but those should have their own specific issues, so I'm closing this one.
